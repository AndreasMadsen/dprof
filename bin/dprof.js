#!/bin/env node

var endpoint = require('endpoint');
var extend = require('util-extend');

process.stdin.pipe(endpoint(function (err, dprof) {
  if (err) throw err;
  dprof = JSON.parse(dprof);

  var content = new Node(dprof.root, dprof.total, 1, new State(), new Root());

  var svg = new Element('svg').attr({ xmlns: 'http://www.w3.org/2000/svg', width: 1000 });

  process.stdout.write(svg.start() + '\n');
  process.stdout.write(content.stringify() + '\n');
  process.stdout.write(svg.end());
}));

function Root() {
  this.top = 0;
}

function State() {
  this.y = 0;
}

function Node(node, total, depth, state, parent) {
  this.name = node.name;
  this.init = Math.floor(node.init / total * 1000);
  this.before = Math.floor((node.init + node.before) / total * 1000);
  this.after = Math.floor((node.init + node.after) / total * 1000);
  this.stack = node.stack;

  this.depth = depth;
  this.state = state;
  this.parent = parent;
  this.top = state.y;

  this.elems = [];
  this.elems.push(new Element('rect').attr({
    x: this.init,
    y: this.top,
    width: this.before - this.init,
    height: 20,
    fill: 'SteelBlue',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('rect').attr({
    x: this.before,
    y: this.top,
    width: this.after - this.before,
    height: 20,
    fill: 'IndianRed',
    style: 'shape-rendering: crispEdges'
  }));
  this.elems.push(new Element('line').attr({
    x1: this.init,
    y1: this.parent.top + 20,
    x2: this.init,
    y2: this.top + 20,
    style: 'stroke: #000; stroke-width: 2px'
  }));

  this.state.y += 20;
  this.children = node.children.map(function (child) {
    return new Node(child, total, depth + 1, state, this);
  }, this);
}

Node.prototype._indent = function () {
  var s = '';
  for (var i = 0; i < this.depth; i++) {
    s += '  ';
  }
  return s;
};

Node.prototype.stringify = function () {
  var content = [];
  this.elems.forEach(function (elem) {
    content.push(this._indent() + elem.singleton());
  }, this);
  this.children.forEach(function (child) {
    content.push(child.stringify());
  }, this);

  return content.join('\n');
};

function Element(name) {
  this._name = name;
  this._attr = {};
}

Element.prototype.attr = function (attr) {
  extend(this._attr, attr);
  return this;
};

Element.prototype.start = function () {
  var start = '<' + this._name;
  Object.keys(this._attr).forEach(function (k) {
    start += ' ' + k + '="' + this._attr[k] + '"';
  }, this);
  start += '>';
  return start;
};

Element.prototype.end = function () {
  return '</' + this._name + '>';
};

Element.prototype.singleton = function () {
  return this.start() + this.end();
};
