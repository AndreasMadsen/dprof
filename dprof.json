{
 "name": "root",
 "init": 0,
 "before": 0,
 "after": 4319266,
 "children": [
  {
   "name": "Timer",
   "init": 2044586,
   "before": 548667,
   "after": 1409916,
   "children": [
    {
     "name": "FSReqWrap",
     "init": 3039113,
     "before": 458225,
     "after": 788615,
     "children": [
      {
       "name": "FSReqWrap",
       "init": 3776193,
       "before": 164991,
       "after": 213630,
       "children": []
      },
      {
       "name": "FSReqWrap",
       "init": 3803875,
       "before": 188243,
       "after": 268189,
       "children": [
        {
         "name": "FSReqWrap",
         "init": 4063866,
         "before": 39850,
         "after": 74510,
         "children": []
        }
       ]
      }
     ]
    },
    {
     "name": "FSReqWrap",
     "init": 3279865,
     "before": 623396,
     "after": 650968,
     "children": []
    }
   ]
  }
 ]
}