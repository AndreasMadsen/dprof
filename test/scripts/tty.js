'use strict';
require('../../dprof.js');

var fs = require('fs');

fs.createReadStream(__filename).pipe(process.stdout);
