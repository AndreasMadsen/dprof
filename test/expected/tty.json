{
 "stdout": "'use strict';\nrequire('../../dprof.js');\n\nconst fs = require('fs');\n\nfs.createReadStream(__filename).pipe(process.stdout);\n",
 "stderr": "init { uid: 3, type: 'SIGNALWRAP', triggerId: 1 }\ninit { uid: 4, type: 'FSREQWRAP', triggerId: 1 }\ninit { uid: 5, type: 'PIPEWRAP', triggerId: 1 }\ninit { uid: 6, type: 'TickObject', triggerId: 1 }\nbefore { uid: 6 }\nafter { uid: 6 }\ndestroy { uid: 6 }\nbefore { uid: 4 }\ninit { uid: 7, type: 'FSREQWRAP', triggerId: 4 }\nafter { uid: 4 }\ndestroy { uid: 4 }\nbefore { uid: 7 }\ninit { uid: 8, type: 'TickObject', triggerId: 7 }\ninit { uid: 9, type: 'FSREQWRAP', triggerId: 7 }\ninit { uid: 10, type: 'TickObject', triggerId: 7 }\nafter { uid: 7 }\nbefore { uid: 8 }\nafter { uid: 8 }\nbefore { uid: 10 }\nafter { uid: 10 }\ndestroy { uid: 8 }\ndestroy { uid: 10 }\ndestroy { uid: 7 }\nbefore { uid: 9 }\ninit { uid: 11, type: 'TickObject', triggerId: 9 }\nafter { uid: 9 }\nbefore { uid: 11 }\ninit { uid: 12, type: 'FSREQWRAP', triggerId: 11 }\nafter { uid: 11 }\ndestroy { uid: 11 }\ndestroy { uid: 9 }\nbefore { uid: 12 }\nafter { uid: 12 }\n",
 "dump": {
  "total": 23000408,
  "version": "1.0.0",
  "root": {
   "name": "root",
   "parent": null,
   "uid": 1,
   "init": 0,
   "destroy": 16392715,
   "before": [
    0
   ],
   "after": [
    16373441
   ],
   "unref": [],
   "ref": [],
   "initRef": true,
   "stack": [
    {
     "description": "Module._compile (module.js:569:30)",
     "filename": "module.js",
     "line": 569,
     "collum": 30
    },
    {
     "description": "Object.Module._extensions..js (module.js:580:10)",
     "filename": "module.js",
     "line": 580,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:503:32)",
     "filename": "module.js",
     "line": 503,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:466:12)",
     "filename": "module.js",
     "line": 466,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:458:3)",
     "filename": "module.js",
     "line": 458,
     "collum": 3
    },
    {
     "description": "Module.require (module.js:513:17)",
     "filename": "module.js",
     "line": 513,
     "collum": 17
    },
    {
     "description": "require (internal/module.js:11:18)",
     "filename": "internal/module.js",
     "line": 11,
     "collum": 18
    },
    {
     "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tty.js:2:1)",
     "filename": "/test/scripts/tty.js",
     "line": 2,
     "collum": 1
    },
    {
     "description": "Module._compile (module.js:569:30)",
     "filename": "module.js",
     "line": 569,
     "collum": 30
    },
    {
     "description": "Object.Module._extensions..js (module.js:580:10)",
     "filename": "module.js",
     "line": 580,
     "collum": 10
    },
    {
     "description": "Module.load (module.js:503:32)",
     "filename": "module.js",
     "line": 503,
     "collum": 32
    },
    {
     "description": "tryModuleLoad (module.js:466:12)",
     "filename": "module.js",
     "line": 466,
     "collum": 12
    },
    {
     "description": "Function.Module._load (module.js:458:3)",
     "filename": "module.js",
     "line": 458,
     "collum": 3
    },
    {
     "description": "Function.Module.runMain (module.js:605:10)",
     "filename": "module.js",
     "line": 605,
     "collum": 10
    },
    {
     "description": "startup (bootstrap_node.js:158:16)",
     "filename": "bootstrap_node.js",
     "line": 158,
     "collum": 16
    },
    {
     "description": "bootstrap_node.js:575:3",
     "filename": "bootstrap_node.js",
     "line": 575,
     "collum": 3
    }
   ],
   "children": [
    3,
    4,
    5,
    6
   ]
  },
  "nodes": [
   {
    "name": "SIGNALWRAP",
    "parent": 1,
    "uid": 3,
    "init": 4330972,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [
     4679844
    ],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "Signal.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "process.<anonymous> (internal/process.js:203:20)",
      "filename": "internal/process.js",
      "line": 203,
      "collum": 20
     },
     {
      "description": "emitTwo (events.js:125:13)",
      "filename": "events.js",
      "line": 125,
      "collum": 13
     },
     {
      "description": "process.emit (events.js:213:7)",
      "filename": "events.js",
      "line": 213,
      "collum": 7
     },
     {
      "description": "_addListener (events.js:248:14)",
      "filename": "events.js",
      "line": 248,
      "collum": 14
     },
     {
      "description": "process.addListener (events.js:298:10)",
      "filename": "events.js",
      "line": 298,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/dprof.js:206:11)",
      "filename": "/dprof.js",
      "line": 206,
      "collum": 11
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Module.require (module.js:513:17)",
      "filename": "module.js",
      "line": 513,
      "collum": 17
     },
     {
      "description": "require (internal/module.js:11:18)",
      "filename": "internal/module.js",
      "line": 11,
      "collum": 18
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tty.js:2:1)",
      "filename": "/test/scripts/tty.js",
      "line": 2,
      "collum": 1
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "FSREQWRAP",
    "parent": 1,
    "uid": 4,
    "init": 5424316,
    "destroy": 18255343,
    "before": [
     17454112
    ],
    "after": [
     18208878
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "FSReqWrap.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "Object.fs.open (fs.js:696:11)",
      "filename": "fs.js",
      "line": 696,
      "collum": 11
     },
     {
      "description": "ReadStream.open (fs.js:1992:6)",
      "filename": "fs.js",
      "line": 1992,
      "collum": 6
     },
     {
      "description": "new ReadStream (fs.js:1979:10)",
      "filename": "fs.js",
      "line": 1979,
      "collum": 10
     },
     {
      "description": "Object.fs.createReadStream (fs.js:1933:10)",
      "filename": "fs.js",
      "line": 1933,
      "collum": 10
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tty.js:6:4)",
      "filename": "/test/scripts/tty.js",
      "line": 6,
      "collum": 4
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": [
     7
    ]
   },
   {
    "name": "PIPEWRAP",
    "parent": 1,
    "uid": 5,
    "init": 14023684,
    "destroy": null,
    "before": [],
    "after": [],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "Pipe.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "createHandle (net.js:58:31)",
      "filename": "net.js",
      "line": 58,
      "collum": 31
     },
     {
      "description": "new Socket (net.js:203:20)",
      "filename": "net.js",
      "line": 203,
      "collum": 20
     },
     {
      "description": "createWritableStdioStream (internal/process/stdio.js:173:16)",
      "filename": "internal/process/stdio.js",
      "line": 173,
      "collum": 16
     },
     {
      "description": "process.getStdout [as stdout] (internal/process/stdio.js:20:14)",
      "filename": "internal/process/stdio.js",
      "line": 20,
      "collum": 14
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tty.js:6:45)",
      "filename": "/test/scripts/tty.js",
      "line": 6,
      "collum": 45
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "TickObject",
    "parent": 1,
    "uid": 6,
    "init": 15620506,
    "destroy": 17366791,
    "before": [
     16630305
    ],
    "after": [
     17097543
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "emitInitS (async_hooks.js:327:7)",
      "filename": "async_hooks.js",
      "line": 327,
      "collum": 7
     },
     {
      "description": "setupInit (internal/process/next_tick.js:225:7)",
      "filename": "internal/process/next_tick.js",
      "line": 225,
      "collum": 7
     },
     {
      "description": "process.nextTick (internal/process/next_tick.js:247:5)",
      "filename": "internal/process/next_tick.js",
      "line": 247,
      "collum": 5
     },
     {
      "description": "resume (_stream_readable.js:802:13)",
      "filename": "_stream_readable.js",
      "line": 802,
      "collum": 13
     },
     {
      "description": "ReadStream.Readable.resume (_stream_readable.js:794:5)",
      "filename": "_stream_readable.js",
      "line": 794,
      "collum": 5
     },
     {
      "description": "ReadStream.Readable.on (_stream_readable.js:764:12)",
      "filename": "_stream_readable.js",
      "line": 764,
      "collum": 12
     },
     {
      "description": "ReadStream.Readable.pipe (_stream_readable.js:622:7)",
      "filename": "_stream_readable.js",
      "line": 622,
      "collum": 7
     },
     {
      "description": "Object.<anonymous> (/Users/Andreas/Sites/node_modules/dprof/test/scripts/tty.js:6:33)",
      "filename": "/test/scripts/tty.js",
      "line": 6,
      "collum": 33
     },
     {
      "description": "Module._compile (module.js:569:30)",
      "filename": "module.js",
      "line": 569,
      "collum": 30
     },
     {
      "description": "Object.Module._extensions..js (module.js:580:10)",
      "filename": "module.js",
      "line": 580,
      "collum": 10
     },
     {
      "description": "Module.load (module.js:503:32)",
      "filename": "module.js",
      "line": 503,
      "collum": 32
     },
     {
      "description": "tryModuleLoad (module.js:466:12)",
      "filename": "module.js",
      "line": 466,
      "collum": 12
     },
     {
      "description": "Function.Module._load (module.js:458:3)",
      "filename": "module.js",
      "line": 458,
      "collum": 3
     },
     {
      "description": "Function.Module.runMain (module.js:605:10)",
      "filename": "module.js",
      "line": 605,
      "collum": 10
     },
     {
      "description": "startup (bootstrap_node.js:158:16)",
      "filename": "bootstrap_node.js",
      "line": 158,
      "collum": 16
     },
     {
      "description": "bootstrap_node.js:575:3",
      "filename": "bootstrap_node.js",
      "line": 575,
      "collum": 3
     }
    ],
    "children": []
   },
   {
    "name": "FSREQWRAP",
    "parent": 4,
    "uid": 7,
    "init": 17977541,
    "destroy": 21201779,
    "before": [
     18322778
    ],
    "after": [
     20539867
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "FSReqWrap.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "Object.fs.read (fs.js:724:11)",
      "filename": "fs.js",
      "line": 724,
      "collum": 11
     },
     {
      "description": "ReadStream._read (fs.js:2039:6)",
      "filename": "fs.js",
      "line": 2039,
      "collum": 6
     },
     {
      "description": "ReadStream.<anonymous> (fs.js:2011:12)",
      "filename": "fs.js",
      "line": 2011,
      "collum": 12
     },
     {
      "description": "Object.onceWrapper (events.js:312:19)",
      "filename": "events.js",
      "line": 312,
      "collum": 19
     },
     {
      "description": "emitOne (events.js:115:13)",
      "filename": "events.js",
      "line": 115,
      "collum": 13
     },
     {
      "description": "ReadStream.emit (events.js:210:7)",
      "filename": "events.js",
      "line": 210,
      "collum": 7
     },
     {
      "description": "fs.js:2002:10",
      "filename": "fs.js",
      "line": 2002,
      "collum": 10
     },
     {
      "description": "FSReqWrap.oncomplete (fs.js:135:15)",
      "filename": "fs.js",
      "line": 135,
      "collum": 15
     }
    ],
    "children": [
     8,
     9,
     10
    ]
   },
   {
    "name": "TickObject",
    "parent": 7,
    "uid": 8,
    "init": 19641630,
    "destroy": 21116412,
    "before": [
     20725051
    ],
    "after": [
     20899764
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "emitInitS (async_hooks.js:327:7)",
      "filename": "async_hooks.js",
      "line": 327,
      "collum": 7
     },
     {
      "description": "setupInit (internal/process/next_tick.js:225:7)",
      "filename": "internal/process/next_tick.js",
      "line": 225,
      "collum": 7
     },
     {
      "description": "process.nextTick (internal/process/next_tick.js:247:5)",
      "filename": "internal/process/next_tick.js",
      "line": 247,
      "collum": 5
     },
     {
      "description": "onwrite (_stream_writable.js:419:15)",
      "filename": "_stream_writable.js",
      "line": 419,
      "collum": 15
     },
     {
      "description": "Socket._writeGeneric (net.js:768:5)",
      "filename": "net.js",
      "line": 768,
      "collum": 5
     },
     {
      "description": "Socket._write (net.js:778:8)",
      "filename": "net.js",
      "line": 778,
      "collum": 8
     },
     {
      "description": "doWrite (_stream_writable.js:371:12)",
      "filename": "_stream_writable.js",
      "line": 371,
      "collum": 12
     },
     {
      "description": "writeOrBuffer (_stream_writable.js:357:5)",
      "filename": "_stream_writable.js",
      "line": 357,
      "collum": 5
     },
     {
      "description": "Socket.Writable.write (_stream_writable.js:274:11)",
      "filename": "_stream_writable.js",
      "line": 274,
      "collum": 11
     },
     {
      "description": "Socket.write (net.js:696:40)",
      "filename": "net.js",
      "line": 696,
      "collum": 40
     },
     {
      "description": "ReadStream.ondata (_stream_readable.js:626:20)",
      "filename": "_stream_readable.js",
      "line": 626,
      "collum": 20
     },
     {
      "description": "emitOne (events.js:115:13)",
      "filename": "events.js",
      "line": 115,
      "collum": 13
     },
     {
      "description": "ReadStream.emit (events.js:210:7)",
      "filename": "events.js",
      "line": 210,
      "collum": 7
     },
     {
      "description": "addChunk (_stream_readable.js:250:12)",
      "filename": "_stream_readable.js",
      "line": 250,
      "collum": 12
     },
     {
      "description": "readableAddChunk (_stream_readable.js:237:11)",
      "filename": "_stream_readable.js",
      "line": 237,
      "collum": 11
     },
     {
      "description": "ReadStream.Readable.push (_stream_readable.js:195:10)",
      "filename": "_stream_readable.js",
      "line": 195,
      "collum": 10
     },
     {
      "description": "onread (fs.js:2059:12)",
      "filename": "fs.js",
      "line": 2059,
      "collum": 12
     },
     {
      "description": "FSReqWrap.wrapper [as oncomplete] (fs.js:718:17)",
      "filename": "fs.js",
      "line": 718,
      "collum": 17
     }
    ],
    "children": []
   },
   {
    "name": "FSREQWRAP",
    "parent": 7,
    "uid": 9,
    "init": 19987136,
    "destroy": 22672231,
    "before": [
     21247418
    ],
    "after": [
     21617137
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "FSReqWrap.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "Object.fs.read (fs.js:724:11)",
      "filename": "fs.js",
      "line": 724,
      "collum": 11
     },
     {
      "description": "ReadStream._read (fs.js:2039:6)",
      "filename": "fs.js",
      "line": 2039,
      "collum": 6
     },
     {
      "description": "ReadStream.Readable.read (_stream_readable.js:429:10)",
      "filename": "_stream_readable.js",
      "line": 429,
      "collum": 10
     },
     {
      "description": "addChunk (_stream_readable.js:251:12)",
      "filename": "_stream_readable.js",
      "line": 251,
      "collum": 12
     },
     {
      "description": "readableAddChunk (_stream_readable.js:237:11)",
      "filename": "_stream_readable.js",
      "line": 237,
      "collum": 11
     },
     {
      "description": "ReadStream.Readable.push (_stream_readable.js:195:10)",
      "filename": "_stream_readable.js",
      "line": 195,
      "collum": 10
     },
     {
      "description": "onread (fs.js:2059:12)",
      "filename": "fs.js",
      "line": 2059,
      "collum": 12
     },
     {
      "description": "FSReqWrap.wrapper [as oncomplete] (fs.js:718:17)",
      "filename": "fs.js",
      "line": 718,
      "collum": 17
     }
    ],
    "children": [
     11
    ]
   },
   {
    "name": "TickObject",
    "parent": 7,
    "uid": 10,
    "init": 20232292,
    "destroy": 21164283,
    "before": [
     20947107
    ],
    "after": [
     21045376
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "emitInitS (async_hooks.js:327:7)",
      "filename": "async_hooks.js",
      "line": 327,
      "collum": 7
     },
     {
      "description": "setupInit (internal/process/next_tick.js:225:7)",
      "filename": "internal/process/next_tick.js",
      "line": 225,
      "collum": 7
     },
     {
      "description": "process.nextTick (internal/process/next_tick.js:247:5)",
      "filename": "internal/process/next_tick.js",
      "line": 247,
      "collum": 5
     },
     {
      "description": "maybeReadMore (_stream_readable.js:514:13)",
      "filename": "_stream_readable.js",
      "line": 514,
      "collum": 13
     },
     {
      "description": "addChunk (_stream_readable.js:263:3)",
      "filename": "_stream_readable.js",
      "line": 263,
      "collum": 3
     },
     {
      "description": "readableAddChunk (_stream_readable.js:237:11)",
      "filename": "_stream_readable.js",
      "line": 237,
      "collum": 11
     },
     {
      "description": "ReadStream.Readable.push (_stream_readable.js:195:10)",
      "filename": "_stream_readable.js",
      "line": 195,
      "collum": 10
     },
     {
      "description": "onread (fs.js:2059:12)",
      "filename": "fs.js",
      "line": 2059,
      "collum": 12
     },
     {
      "description": "FSReqWrap.wrapper [as oncomplete] (fs.js:718:17)",
      "filename": "fs.js",
      "line": 718,
      "collum": 17
     }
    ],
    "children": []
   },
   {
    "name": "TickObject",
    "parent": 9,
    "uid": 11,
    "init": 21508840,
    "destroy": 22640026,
    "before": [
     21658968
    ],
    "after": [
     22585497
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "emitInitS (async_hooks.js:327:7)",
      "filename": "async_hooks.js",
      "line": 327,
      "collum": 7
     },
     {
      "description": "setupInit (internal/process/next_tick.js:225:7)",
      "filename": "internal/process/next_tick.js",
      "line": 225,
      "collum": 7
     },
     {
      "description": "process.nextTick (internal/process/next_tick.js:247:5)",
      "filename": "internal/process/next_tick.js",
      "line": 247,
      "collum": 5
     },
     {
      "description": "endReadable (_stream_readable.js:1036:13)",
      "filename": "_stream_readable.js",
      "line": 1036,
      "collum": 13
     },
     {
      "description": "ReadStream.Readable.read (_stream_readable.js:380:7)",
      "filename": "_stream_readable.js",
      "line": 380,
      "collum": 7
     },
     {
      "description": "flow (_stream_readable.js:833:34)",
      "filename": "_stream_readable.js",
      "line": 833,
      "collum": 34
     },
     {
      "description": "emitReadable_ (_stream_readable.js:501:3)",
      "filename": "_stream_readable.js",
      "line": 501,
      "collum": 3
     },
     {
      "description": "emitReadable (_stream_readable.js:494:7)",
      "filename": "_stream_readable.js",
      "line": 494,
      "collum": 7
     },
     {
      "description": "onEofChunk (_stream_readable.js:479:3)",
      "filename": "_stream_readable.js",
      "line": 479,
      "collum": 3
     },
     {
      "description": "readableAddChunk (_stream_readable.js:207:5)",
      "filename": "_stream_readable.js",
      "line": 207,
      "collum": 5
     },
     {
      "description": "ReadStream.Readable.push (_stream_readable.js:195:10)",
      "filename": "_stream_readable.js",
      "line": 195,
      "collum": 10
     },
     {
      "description": "onread (fs.js:2059:12)",
      "filename": "fs.js",
      "line": 2059,
      "collum": 12
     },
     {
      "description": "FSReqWrap.wrapper [as oncomplete] (fs.js:718:17)",
      "filename": "fs.js",
      "line": 718,
      "collum": 17
     }
    ],
    "children": [
     12
    ]
   },
   {
    "name": "FSREQWRAP",
    "parent": 11,
    "uid": 12,
    "init": 22096110,
    "destroy": null,
    "before": [
     22728583
    ],
    "after": [
     22869140
    ],
    "unref": [],
    "ref": [],
    "initRef": true,
    "stack": [
     {
      "description": "runInitCallback (async_hooks.js:459:5)",
      "filename": "async_hooks.js",
      "line": 459,
      "collum": 5
     },
     {
      "description": "FSReqWrap.init (async_hooks.js:441:7)",
      "filename": "async_hooks.js",
      "line": 441,
      "collum": 7
     },
     {
      "description": "Object.fs.close (fs.js:668:11)",
      "filename": "fs.js",
      "line": 668,
      "collum": 11
     },
     {
      "description": "ReadStream.close (fs.js:2086:6)",
      "filename": "fs.js",
      "line": 2086,
      "collum": 6
     },
     {
      "description": "ReadStream._destroy (fs.js:2066:8)",
      "filename": "fs.js",
      "line": 2066,
      "collum": 8
     },
     {
      "description": "ReadStream.destroy (internal/streams/destroy.js:32:8)",
      "filename": "internal/streams/destroy.js",
      "line": 32,
      "collum": 8
     },
     {
      "description": "ReadStream.<anonymous> (fs.js:1983:12)",
      "filename": "fs.js",
      "line": 1983,
      "collum": 12
     },
     {
      "description": "emitNone (events.js:110:20)",
      "filename": "events.js",
      "line": 110,
      "collum": 20
     },
     {
      "description": "ReadStream.emit (events.js:207:7)",
      "filename": "events.js",
      "line": 207,
      "collum": 7
     },
     {
      "description": "endReadableNT (_stream_readable.js:1045:12)",
      "filename": "_stream_readable.js",
      "line": 1045,
      "collum": 12
     },
     {
      "description": "_combinedTickCallback (internal/process/next_tick.js:102:11)",
      "filename": "internal/process/next_tick.js",
      "line": 102,
      "collum": 11
     },
     {
      "description": "process._tickCallback (internal/process/next_tick.js:161:9)",
      "filename": "internal/process/next_tick.js",
      "line": 161,
      "collum": 9
     }
    ],
    "children": []
   }
  ]
 }
}